src/mpi/comm/comm_split_type.c

```diff
@@ -152,14 +152,52 @@ int MPIR_Comm_split_type_hw_guided(MPIR_Comm * comm_ptr, int key,
     goto fn_exit;
 }
 
-int MPIR_Comm_split_type_hw_unguided(MPIR_Comm * user_comm_ptr, int key,
-                                     MPIR_Info * info_ptr, MPIR_Comm ** newcomm_ptr)
+int MPIR_Comm_split_type_hw_unguided(MPIR_Comm * comm_ptr, int key, MPIR_Comm ** newcomm_ptr)
 {
     int mpi_errno = MPI_SUCCESS;
+    MPIR_Comm *subcomm = NULL;
 
+    int orig_size = MPIR_Comm_size(comm_ptr);
+
+    /* we'll try top-down until we hit upon a subset */
+    mpi_errno = MPIR_Comm_split_type_by_node(comm_ptr, key, &subcomm);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    if (MPIR_Comm_size(subcomm) < orig_size) {
+        *newcomm_ptr = subcomm;
+        goto fn_exit;
+    }
+
+    /* TODO: determine the "proper" hierarchy */
+    const char *topolist[] = {
+        "socket",
+        "package",
+        "numanode",
+        "l3cache",
+        "l2cache",
+        "l1cache",
+        "cpu",
+        "core",
+        "hwthread",
+    };
+
+    for (int i = 0; i < sizeof(topolist) / sizeof(topolist[0]); i++) {
+        mpi_errno = node_split(comm_ptr, key, topolist[i], &subcomm);
+        MPIR_ERR_CHECK(mpi_errno);
+
+        if (MPIR_Comm_size(subcomm) < orig_size) {
```

Comment:
Maybe it is good to add a note that we return the first successful subcomm

Reply:
Comment added

Reply:
I don't see comment added here in commit "comm: implment MPIR_Comm_split_type_hw_unguided". Did you add into a separate commit?

Reply:
I thought the comment at the beginning of the function is sufficient: https://github.com/pmodels/mpich/blob/881ef456c47bac6a34ddce03e8d571104e95bb41/src/mpi/comm/comm_split_type.c#L165-L170

Let me add another short one here.